### PR TITLE
refine docs for regexp tree dictionary

### PR DESCRIPTION
**Pull request by hanfei1991**
### Changelog category (leave one):

- Documentation (changelog entry is not required)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
...

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

**Comment by robot-ch-test-poll** [comment]
This is an automated comment for commit 29aa96037758df616ec6253289e218abd7212626 with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/49902/29aa96037758df616ec6253289e218abd7212626/ci_running.html)
The overall status of the commit is 🟢 success

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>Style Check</td><td>Runs a set of checks to keep the code style clean. If some of tests failed, see the related log from the report</td><td>🟢 success</td></tr>
</table>

File: docs/en/sql-reference/dictionaries/index.md [hunk 1]
```diff
@@ -2197,13 +2197,13 @@ Result:
 └─────────────────────────────────┴───────┘
 ```
 
-## RegExp Tree Dictionary {#regexp-tree-dictionary}
+## RegExpTree Dictionary {#regexp-tree-dictionary}
 
-Regexp Tree dictionary stores multiple trees of regular expressions with attributions. Users can retrieve strings in the dictionary. If a string matches the root of the regexp tree, we will collect the corresponding attributes of the matched root and continue to walk the children. If any of the children matches the string, we will collect attributes and rewrite the old ones if conflicts occur, then continue the traverse until we reach leaf nodes.
+RegExpTree Dictionary is designed to store multiple regular expressions in a dictionary, and query if a string could match one or multiple regular expressions. In some senarioes, for example, User Agent Parser, this data structure is very useful. We can use it in both local and cloud environment.
```
Comment (serxa) [inline docs/en/sql-reference/dictionaries/index.md:2202]:
```suggestion
RegExpTree Dictionary is designed to store multiple regular expressions in a dictionary, and query if a string could match one or multiple regular expressions. In some scenarios, for example, User Agent Parser, this data structure is very useful. We can use it in both local and cloud environments.
```

File: docs/en/sql-reference/dictionaries/index.md [hunk 2]
```diff
@@ -2260,12 +2256,87 @@ SELECT dictGet('regexp_dict', ('name', 'version'), '31/tclwebkit1024');
...
 └─────────────────────────────────────────────────────────────────┘
 ```
 
+Explain:
+
+In this case, we match the regular expression `\d+/tclwebkit(?:\d+[\.\d]*)` in the first layer, so the dictionary will continue to look into the children nodes in the second layer and find it matches `3[12]/tclwebkit`. As a result, the value of `name` is `Andriod` defined in the first layer and the value of `version` is `12` defined in the second layer.
+
+### Use RegExpTree Dictionary on cloud
+
+We have shown how RegExpTree work in the local enviroument, but we cannot use `YAMLRegExpTree` on cloud. If we have a local yaml file, we can use this file to create RegExpTree Dictionary in the local enviroment, then dump this dictionary to a csv file by `dictionary` table function and [INTO OUTFILE](../../statements/select/into-outfile.md) clause.
```
Comment (serxa) [inline docs/en/sql-reference/dictionaries/index.md:2271]:
```suggestion
We have shown how RegExpTree work in the local environment, but we cannot use `YAMLRegExpTree` in the cloud. If we have a local yaml file, we can use this file to create RegExpTree Dictionary in the local environment, then dump this dictionary to a csv file by the `dictionary` table function and [INTO OUTFILE](../../statements/select/into-outfile.md) clause.
```

File: docs/en/sql-reference/dictionaries/index.md [hunk 3]
```diff
@@ -2260,12 +2256,87 @@ SELECT dictGet('regexp_dict', ('name', 'version'), '31/tclwebkit1024');
...
+    version String
+PRIMARY KEY(regexp)
+SOURCE(CLICKHOUSE(TABLE 'regexp_dictionary_source_table'))
+LIFETIME(0)
+LAYOUT(regexp_tree);
+```
+
+### Use RegexpTree Dictionary as a UA Parser
+
+With a powerful yaml configure file, we can use RepexpTree dictionary as a UA parser. We support [uap-core](https://github.com/ua-parser/uap-core) and demostrate how to use it in the functional test [02504_regexp_dictionary_ua_parser](https://github.com/ClickHouse/ClickHouse/blob/master/tests/queries/0_stateless/02504_regexp_dictionary_ua_parser.sh)
```
Comment (serxa) [inline docs/en/sql-reference/dictionaries/index.md:2338]:
```suggestion
With a powerful yaml configure file, we can use RepexpTree dictionary as a UA parser. We support [uap-core](https://github.com/ua-parser/uap-core) and demonstrate how to use it in the functional test [02504_regexp_dictionary_ua_parser](https://github.com/ClickHouse/ClickHouse/blob/master/tests/queries/0_stateless/02504_regexp_dictionary_ua_parser.sh)
```

**Review by serxa — APPROVED** [review]

File: docs/en/sql-reference/dictionaries/index.md [hunk 4]
```diff
@@ -2240,17 +2238,15 @@ We introduce a type of source called `YAMLRegExpTree` representing the structure
...
-**Back Reference**
-
-The value of an attribution could contain a back reference which refers to a capture group of the matched regular expression. Reference number ranges from 1 to 9 and writes as `$1` or `\1`.
-
-During the query execution, the back reference in the value will be replaced by the matched capture group.
+This config consists of a list of RegExpTree nodes. Each node has following structure:
 
-**Query**
+- **regexp** means the regular expression of this node.
+- **user defined attributions** is a list of dictionary attributions defined in the dictionary structure. In this case, we have two attributions: `name` and `version`. The first nodes have both attributions. The second node only has `name` attribution, because the `version` is defined in the children nodes.
```
Comment (rschu1ze) [inline docs/en/sql-reference/dictionaries/index.md:2244]: The rest of the document talks about "attributes". The term "attribution" occurs nowhere else. Could we just use "attributes"?

File: docs/en/sql-reference/dictionaries/index.md [hunk 5]
```diff
@@ -2260,12 +2256,87 @@ SELECT dictGet('regexp_dict', ('name', 'version'), '31/tclwebkit1024');
...
-```
+```text
 ┌─dictGet('regexp_dict', ('name', 'version'), '31/tclwebkit1024')─┐
 │ ('Andriod','12')                                                │
 └─────────────────────────────────────────────────────────────────┘
 ```
 
+Explain:
+
+In this case, we match the regular expression `\d+/tclwebkit(?:\d+[\.\d]*)` in the first layer, so the dictionary will continue to look into the children nodes in the second layer and find it matches `3[12]/tclwebkit`. As a result, the value of `name` is `Andriod` defined in the first layer and the value of `version` is `12` defined in the second layer.
```
Comment (rschu1ze) [inline docs/en/sql-reference/dictionaries/index.md:2267]: "in the first layer" --> "in the top layer's second node"

Comment (rschu1ze) [reply docs/en/sql-reference/dictionaries/index.md:2267]: Wrong (?) spelling "Andriod" pops up again.

File: docs/en/sql-reference/dictionaries/index.md [hunk 6]
```diff
@@ -2260,12 +2256,87 @@ SELECT dictGet('regexp_dict', ('name', 'version'), '31/tclwebkit1024');
...
+
+Then update the local CSV by
+
+```bash
+clickhouse client \
+    --host MY_HOST \
+    --secure \
+    --password MY_PASSWORD \
+    --query "
+    insert into regexp_dictionary_source_table
```
Comment (rschu1ze) [inline docs/en/sql-reference/dictionaries/index.md:2317]: Cosmetics: Let's write all SQL keywords (INSERT INTO, SELECT, etc.) UPPERCASE.

File: docs/en/sql-reference/dictionaries/index.md [hunk 7]
```diff
@@ -2260,12 +2256,87 @@ SELECT dictGet('regexp_dict', ('name', 'version'), '31/tclwebkit1024');
...
+    version String
+PRIMARY KEY(regexp)
+SOURCE(CLICKHOUSE(TABLE 'regexp_dictionary_source_table'))
+LIFETIME(0)
+LAYOUT(regexp_tree);
+```
+
+### Use RegexpTree Dictionary as a UA Parser
+
+With a powerful yaml configure file, we can use RepexpTree dictionary as a UA parser. We support [uap-core](https://github.com/ua-parser/uap-core) and demonstrate how to use it in the functional test [02504_regexp_dictionary_ua_parser](https://github.com/ClickHouse/ClickHouse/blob/master/tests/queries/0_stateless/02504_regexp_dictionary_ua_parser.sh)
```
Comment (rschu1ze) [inline docs/en/sql-reference/dictionaries/index.md:2338]: I'd remove section l. 2336ff and move this sentence after l. 2268.

File: docs/en/sql-reference/dictionaries/index.md [hunk 8]
```diff
@@ -2218,11 +2218,9 @@ LAYOUT(regexp_tree)
 ...
 ```
 
-**Source**
-
-We introduce a type of source called `YAMLRegExpTree` representing the structure of Regexp Tree dictionary. An Example of a valid yaml config is like:
+The dictionary source `YAMLRegExpTree` represents the structure of a Regexp Tree. For example:
```
[the remark this answers is not in the document]
Comment (hanfei1991) [reply docs/en/sql-reference/dictionaries/index.md:2221]: no, we also support Clickhouse Table Source as well. Even in local version.

File: docs/en/sql-reference/dictionaries/index.md [hunk 9]
```diff
@@ -2240,17 +2238,15 @@ We introduce a type of source called `YAMLRegExpTree` representing the structure
...
 ```
 
-The key `regexp` represents the regular expression of a tree node. The name of key is same as the dictionary key. The `name` and `version` is user-defined attributions in the dicitionary. The `versions` (which can be any name that not appear in attributions or the key) indicates the children nodes of this tree.
-
-**Back Reference**
-
-The value of an attribution could contain a back reference which refers to a capture group of the matched regular expression. Reference number ranges from 1 to 9 and writes as `$1` or `\1`.
-
-During the query execution, the back reference in the value will be replaced by the matched capture group.
+This config consists of a list of RegExpTree nodes. Each node has following structure:
```
[the remark this answers is not in the document]
Comment (hanfei1991) [reply docs/en/sql-reference/dictionaries/index.md:2241]: hmmmmmmmmm.... it's a typo :D

File: docs/en/sql-reference/dictionaries/index.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (hanfei1991) [reply docs/en/sql-reference/dictionaries/index.md:2244]: The answer is: `versions` is not an attribute, it is `children nodes`. Attributes must appear in dictionary defination. The Yaml node could have:

- regexp
- attribute(s) (`name` and `version` in this case)
- children nodes (`versions` in this case)

I know it's confusting, but I have tried my best. Any suggestions?

File: docs/en/sql-reference/dictionaries/index.md [hunk 10]
```diff
@@ -2240,17 +2238,15 @@ We introduce a type of source called `YAMLRegExpTree` representing the structure
...
 
-**Query**
+- **regexp** means the regular expression of this node.
+- **user defined attributions** is a list of dictionary attributions defined in the dictionary structure. In this case, we have two attributions: `name` and `version`. The first nodes have both attributions. The second node only has `name` attribution, because the `version` is defined in the children nodes.
+  - The value of an attribution could contain a **back reference** which refers to a capture group of the matched regular expression. Reference number ranges from 1 to 9 and writes as `$1` or `\1`. During the query execution, the back reference in the value will be replaced by the matched capture group.
```
[the remark this answers is not in the document]
Comment (hanfei1991) [reply docs/en/sql-reference/dictionaries/index.md:2245]: Right now it's now allowed. I wonder if users need it.

[the remark this answers is not in the document]
Comment (rschu1ze) [reply docs/en/sql-reference/dictionaries/index.md]: Probably not :-)

File: docs/en/sql-reference/dictionaries/index.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (rschu1ze) [reply docs/en/sql-reference/dictionaries/index.md]: I rewrote the docs a bit. Should be fine now.

**Review by rschu1ze — APPROVED** [review]

**Pull request merged**